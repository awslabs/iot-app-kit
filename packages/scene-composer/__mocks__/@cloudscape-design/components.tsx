import React from 'react';
import mockComponent from '../mockComponent';

const cloudscapeRaw: any = jest.createMockFromModule('@cloudscape-design/components');

const cloudscape = Object.keys(cloudscapeRaw).reduce((acc, comp) => {
  if (!comp.startsWith('_')) {
    acc[comp] = mockComponent(comp);
  } else {
    acc[comp] = cloudscapeRaw[comp];
  }
  return acc;
}, {});

module.exports = {
  ...cloudscape,
  ButtonDropdown: ({ items = [], onItemClick, ...props }: any) => {
    return (
      <ul data-mocked='ButtonDropdown' {...props}>
        { items.map(({ id, text, iconSvg, ...itemProps }) => {
          return (<li id={id} {...itemProps} onClick={() => onItemClick({ detail: { id }})}>{iconSvg}{text}</li>)
        })}
      </ul>
    )
  }
};
