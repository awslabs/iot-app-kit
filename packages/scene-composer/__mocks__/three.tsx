/**
 * Mock implementation of R3F (React Three Fiber) components with the string name for a clean snapshot test.
 */
module.exports = {
  ...jest.requireActual('three'),
  audioListener: 'audioListener',
  positionalAudio: 'positionalAudio',

  mesh: 'mesh',
  instancedMesh: 'instancedMesh',
  scene: 'scene',
  sprite: 'sprite',
  lOD: 'lOD',
  skinnedMesh: 'skinnedMesh',
  skeleton: 'skeleton',
  bone: 'bone',
  lineSegments: 'lineSegments',
  lineLoop: 'lineLoop',
  points: 'points',
  group: 'group',
  immediateRenderObject: 'immediateRenderObject',

  // cameras
  camera: 'camera',
  perspectiveCamera: 'perspectiveCamera',
  orthographicCamera: 'orthographicCamera',
  cubeCamera: 'cubeCamera',
  arrayCamera: 'arrayCamera',

  // geometry
  instancedBufferGeometry: 'instancedBufferGeometry',
  bufferGeometry: 'bufferGeometry',
  boxBufferGeometry: 'boxBufferGeometry',
  circleBufferGeometry: 'circleBufferGeometry',
  coneBufferGeometry: 'coneBufferGeometry',
  cylinderBufferGeometry: 'cylinderBufferGeometry',
  dodecahedronBufferGeometry: 'dodecahedronBufferGeometry',
  extrudeBufferGeometry: 'extrudeBufferGeometry',
  icosahedronBufferGeometry: 'icosahedronBufferGeometry',
  latheBufferGeometry: 'latheBufferGeometry',
  octahedronBufferGeometry: 'octahedronBufferGeometry',
  parametricBufferGeometry: 'parametricBufferGeometry',
  planeBufferGeometry: 'planeBufferGeometry',
  polyhedronBufferGeometry: 'polyhedronBufferGeometry',
  ringBufferGeometry: 'ringBufferGeometry',
  shapeBufferGeometry: 'shapeBufferGeometry',
  sphereBufferGeometry: 'sphereBufferGeometry',
  tetrahedronBufferGeometry: 'tetrahedronBufferGeometry',
  textBufferGeometry: 'textBufferGeometry',
  torusBufferGeometry: 'torusBufferGeometry',
  torusKnotBufferGeometry: 'torusKnotBufferGeometry',
  tubeBufferGeometry: 'tubeBufferGeometry',
  wireframeGeometry: 'wireframeGeometry',
  parametricGeometry: 'parametricGeometry',
  tetrahedronGeometry: 'tetrahedronGeometry',
  octahedronGeometry: 'octahedronGeometry',
  icosahedronGeometry: 'icosahedronGeometry',
  dodecahedronGeometry: 'dodecahedronGeometry',
  polyhedronGeometry: 'polyhedronGeometry',
  tubeGeometry: 'tubeGeometry',
  torusKnotGeometry: 'torusKnotGeometry',
  torusGeometry: 'torusGeometry',
  textGeometry: 'textGeometry',
  sphereGeometry: 'sphereGeometry',
  ringGeometry: 'ringGeometry',
  planeGeometry: 'planeGeometry',
  latheGeometry: 'latheGeometry',
  shapeGeometry: 'shapeGeometry',
  extrudeGeometry: 'extrudeGeometry',
  edgesGeometry: 'edgesGeometry',
  coneGeometry: 'coneGeometry',
  cylinderGeometry: 'cylinderGeometry',
  circleGeometry: 'circleGeometry',
  boxGeometry: 'boxGeometry',

  // materials
  material: 'material',
  shadowMaterial: 'shadowMaterial',
  spriteMaterial: 'spriteMaterial',
  rawShaderMaterial: 'rawShaderMaterial',
  shaderMaterial: 'shaderMaterial',
  pointsMaterial: 'pointsMaterial',
  meshPhysicalMaterial: 'meshPhysicalMaterial',
  meshStandardMaterial: 'meshStandardMaterial',
  meshPhongMaterial: 'meshPhongMaterial',
  meshToonMaterial: 'meshToonMaterial',
  meshNormalMaterial: 'meshNormalMaterial',
  meshLambertMaterial: 'meshLambertMaterial',
  meshDepthMaterial: 'meshDepthMaterial',
  meshDistanceMaterial: 'meshDistanceMaterial',
  meshBasicMaterial: 'meshBasicMaterial',
  meshMatcapMaterial: 'meshMatcapMaterial',
  lineDashedMaterial: 'lineDashedMaterial',
  lineBasicMaterial: 'lineBasicMaterial',

  // primitive
  primitive: 'primitive',

  // lights and other
  light: 'light',
  spotLightShadow: 'spotLightShadow',
  spotLight: 'spotLight',
  pointLight: 'pointLight',
  rectAreaLight: 'rectAreaLight',
  hemisphereLight: 'hemisphereLight',
  directionalLightShadow: 'directionalLightShadow',
  directionalLight: 'directionalLight',
  ambientLight: 'ambientLight',
  lightShadow: 'lightShadow',
  ambientLightProbe: 'ambientLightProbe',
  hemisphereLightProbe: 'hemisphereLightProbe',
  lightProbe: 'lightProbe',

  // helpers
  spotLightHelper: 'spotLightHelper',
  skeletonHelper: 'skeletonHelper',
  pointLightHelper: 'pointLightHelper',
  hemisphereLightHelper: 'hemisphereLightHelper',
  gridHelper: 'gridHelper',
  polarGridHelper: 'polarGridHelper',
  directionalLightHelper: 'directionalLightHelper',
  cameraHelper: 'cameraHelper',
  boxHelper: 'boxHelper',
  box3Helper: 'box3Helper',
  planeHelper: 'planeHelper',
  arrowHelper: 'arrowHelper',
  axesHelper: 'axesHelper',

  // textures
  texture: 'texture',
  videoTexture: 'videoTexture',
  dataTexture: 'dataTexture',
  dataTexture3D: 'dataTexture3D',
  compressedTexture: 'compressedTexture',
  cubeTexture: 'cubeTexture',
  canvasTexture: 'canvasTexture',
  depthTexture: 'depthTexture',

  // misc
  raycaster: 'raycaster',
  vector2: 'vector2',
  vector3: 'vector3',
  vector4: 'vector4',
  euler: 'euler',
  matrix3: 'matrix3',
  matrix4: 'matrix4',
  quaternion: 'quaternion',
  bufferAttribute: 'bufferAttribute',
  instancedBufferAttribute: 'instancedBufferAttribute',
  color: 'color',
  fog: 'fog',
  fogExp2: 'fogExp2',
  shape: 'shape',
};
