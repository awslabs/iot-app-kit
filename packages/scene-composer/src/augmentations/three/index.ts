export * from './Anchor';
export * from './Viewpoint';
