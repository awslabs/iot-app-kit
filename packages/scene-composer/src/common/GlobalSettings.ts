import { TwinMakerSceneMetadataModule } from '@iot-app-kit/source-iottwinmaker';
import { MpSdk } from '@matterport/r3f/dist';

import { DracoDecoderConfig, GetSceneObjectFunction } from '../interfaces/sceneViewer';
import { COMPOSER_FEATURES, FeatureConfig } from '../interfaces';
import { IMetricRecorder } from '../interfaces/metricRecorder';

const globalSettings: {
  debugMode: boolean;
  dracoDecoder: DracoDecoderConfig;
  locale: string;
  metricRecorder?: IMetricRecorder;
  featureConfig: FeatureConfig;
  getSceneObjectFunction: GetSceneObjectFunction | undefined;
  twinMakerSceneMetadataModule: TwinMakerSceneMetadataModule | undefined;
  matterportSdks: Record<string, MpSdk>;
} = {
  debugMode: false,
  dracoDecoder: { enable: true },
  locale: 'en-US',
  metricRecorder: undefined,
  // default to disable all features
  featureConfig: {},
  getSceneObjectFunction: undefined,
  twinMakerSceneMetadataModule: undefined,
  matterportSdks: {},
};

const changeSubscribers = [] as Function[];

const notifySubscribers = () => {
  changeSubscribers.forEach((sub) => sub());
};

export const setDebugMode = () => {
  globalSettings.debugMode = true;
  notifySubscribers();
};

export const setDracoDecoder = (dracoDecoder: DracoDecoderConfig) => {
  globalSettings.dracoDecoder = dracoDecoder;
  notifySubscribers();
};
export const setLocale = (locale: string) => {
  globalSettings.locale = locale;
  notifySubscribers();
};

export const setMetricRecorder = (metricRecorder: IMetricRecorder) => {
  globalSettings.metricRecorder = metricRecorder;
  notifySubscribers();
};

export const setFeatureConfig = (featureConfig: Partial<Record<COMPOSER_FEATURES, boolean>>) => {
  globalSettings.featureConfig = featureConfig;
  notifySubscribers();
};

export const setGetSceneObjectFunction = (getSceneObjectFunction: GetSceneObjectFunction) => {
  globalSettings.getSceneObjectFunction = getSceneObjectFunction;
  notifySubscribers();
};

export const setTwinMakerSceneMetadataModule = (twinMakerSceneMetadataModule: TwinMakerSceneMetadataModule) => {
  globalSettings.twinMakerSceneMetadataModule = twinMakerSceneMetadataModule;
  notifySubscribers();
};

export const setMatterportSdk = (sceneId: string, sdk: MpSdk): void => {
  globalSettings.matterportSdks[sceneId] = sdk;
  notifySubscribers();
};

export const getMatterportSdk = (sceneId: string): MpSdk | undefined => {
  return globalSettings.matterportSdks[sceneId];
};

export const getGlobalSettings = () => {
  return globalSettings;
};

export const subscribe = (callback: Function) => {
  changeSubscribers.push(callback);
};

export const unsubscribe = (callback: Function) => {
  changeSubscribers.splice(changeSubscribers.indexOf(callback), 1);
};
