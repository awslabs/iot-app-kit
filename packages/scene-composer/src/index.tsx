export * from './interfaces';

export * from './common/errors';

export * from './components/SceneComposerInternal';

export * from './SceneViewer';

export type { Scene } from './models/SceneModels';

// Augmentation related interfaces
export * from './augmentations/three';

export * from './common/GlobalSettings';

export * from './utils/dataFrameLabelUtils';

export * from './utils/dataBindingTemplateUtils';

export * from './common/constants';
