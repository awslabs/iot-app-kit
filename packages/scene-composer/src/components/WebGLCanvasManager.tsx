import * as THREE from 'three';
import * as awsui from '@awsui/design-tokens';
import React, { useContext, useEffect, useRef, useState } from 'react';
import { GizmoHelper, GizmoViewport } from '@react-three/drei';
import { ThreeEvent, useThree } from '@react-three/fiber';

import { KnownSceneProperty } from '../interfaces';
import useLifecycleLogging from '../logger/react-logger/hooks/useLifecycleLogging';
import { useEditorState, useSceneDocument, useStore } from '../store';
import { sceneComposerIdContext } from '../common/sceneComposerIdContext';
import { hexColorFromDesignToken } from '../utils/styleUtils';
import { Layers, ROOT_OBJECT_3D_NAME } from '../common/constants';
import { getGlobalSettings } from '../common/GlobalSettings';
import { ViewCursorWidget } from '../augmentations/components/three-fiber/viewpoint/ViewCursorWidget';
import { getIntersectionTransform } from '../utils/raycastUtils';
import { createNodeWithPositionAndNormal } from '../utils/nodeUtils';
import { EnvironmentLoadingManager } from '../common/loadingManagers';

import Environment, { presets } from './three-fiber/Environment';
import { StatsWindow } from './three-fiber/StatsWindow';
import EntityGroup from './three-fiber/EntityGroup';
import { EditorMainCamera } from './three-fiber/EditorCamera';
import { EditorTransformControls } from './three-fiber/EditorTransformControls';
import { SceneInfoView } from './three-fiber/SceneInfoView';
import IntlProvider from './IntlProvider';

const GIZMO_MARGIN: [number, number] = [72, 72];

const envLoaderExtension = (loader: THREE.Loader) => {
  loader.manager = EnvironmentLoadingManager;
};

export const WebGLCanvasManager: React.FC = () => {
  const log = useLifecycleLogging('WebGLCanvasManager');

  const sceneComposerId = useContext(sceneComposerIdContext);
  const { isEditing, addingWidget, setAddingWidget } = useEditorState(sceneComposerId);
  const { document, getSceneNodeByRef, getSceneProperty } = useSceneDocument(sceneComposerId);
  const appendSceneNode = useStore(sceneComposerId)((state) => state.appendSceneNode);
  const { gl } = useThree();
  const domRef = useRef<HTMLElement>(gl.domElement.parentElement);
  const environmentPreset = getSceneProperty(KnownSceneProperty.EnvironmentPreset);
  const matterportModelId = getSceneProperty(KnownSceneProperty.MatterportModelId);
  const rootNodeRefs = document.rootNodeRefs;
  const [startingPointerPosition, setStartingPointerPosition] = useState<THREE.Vector2>(new THREE.Vector2());

  const editingTargetPlaneRef = useRef(null);
  const gridHelperRef = useRef<THREE.GridHelper>(null);

  const MAX_CLICK_DISTANCE = 2;

  useEffect(() => {
    if (!!environmentPreset && !(environmentPreset in presets)) {
      log?.error('Environment preset must be one of: ' + Object.keys(presets).join(', '));
    }
  }, [environmentPreset]);

  const onPointerDown = (e: ThreeEvent<PointerEvent>) => {
    setStartingPointerPosition(new THREE.Vector2(e.screenX, e.screenY));
  };

  const onPointerUp = (e: ThreeEvent<MouseEvent>) => {
    const currentPosition = new THREE.Vector2(e.screenX, e.screenY);
    if (startingPointerPosition.distanceTo(currentPosition) <= MAX_CLICK_DISTANCE) {
      if (addingWidget && e.intersections.length > 0) {
        const { position } = getIntersectionTransform(e.intersections[0]);
        const newWidgetNode = createNodeWithPositionAndNormal(addingWidget, position, new THREE.Vector3());

        setAddingWidget(undefined);
        appendSceneNode(newWidgetNode);

        e.stopPropagation();
      }
    }
  };

  useEffect(() => {
    const gridHelper = gridHelperRef.current;
    if (gridHelper) {
      gridHelper.layers.disable(Layers.RaycastAndRender);
      gridHelper.layers.enable(Layers.RenderOnly);
    }
  }, [gridHelperRef.current]);

  return (
    <React.Fragment>
      <EditorMainCamera />
      {environmentPreset in presets && <Environment preset={environmentPreset} extensions={envLoaderExtension} />}
      <group name={ROOT_OBJECT_3D_NAME} dispose={null}>
        {rootNodeRefs &&
          rootNodeRefs.map((rootNodeRef) => {
            const node = getSceneNodeByRef(rootNodeRef);
            return node && <EntityGroup key={rootNodeRef} node={node} />;
          })}
      </group>
      {isEditing() && (
        <React.Fragment>
          <EditorTransformControls />
          <GizmoHelper alignment='bottom-right' margin={GIZMO_MARGIN} renderPriority={0}>
            <GizmoViewport
              axisColors={[
                hexColorFromDesignToken(awsui.colorBackgroundNotificationRed),
                hexColorFromDesignToken(awsui.colorBackgroundNotificationGreen),
                hexColorFromDesignToken(awsui.colorBackgroundNotificationBlue),
              ]}
            />
          </GizmoHelper>
          <ViewCursorWidget />
          <React.Fragment>
            <gridHelper
              ref={gridHelperRef}
              args={[
                1000 /* size */,
                500 /* grid# */,
                new THREE.Color(hexColorFromDesignToken(awsui.colorTextInputPlaceholder)) /* center line color */,
                new THREE.Color(hexColorFromDesignToken(awsui.colorBorderContainerTop)) /* grid color */,
              ]}
            />
            <mesh
              ref={editingTargetPlaneRef}
              name='Ground'
              rotation={[THREE.MathUtils.degToRad(270), 0, 0]}
              onPointerUp={onPointerUp}
              onPointerDown={onPointerDown}
              renderOrder={matterportModelId ? 1 : undefined}
            >
              <planeGeometry args={[1000, 1000]} />
              <meshBasicMaterial colorWrite={false} />
            </mesh>
          </React.Fragment>
          <IntlProvider locale={getGlobalSettings().locale}>
            <SceneInfoView />
          </IntlProvider>

          {getGlobalSettings().debugMode && <StatsWindow parent={domRef} />}
        </React.Fragment>
      )}
    </React.Fragment>
  );
};
