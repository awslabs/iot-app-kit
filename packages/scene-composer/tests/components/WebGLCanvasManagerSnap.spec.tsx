/* eslint-disable import/first */
import renderer from 'react-test-renderer';
import React from 'react';
import { useThree } from '@react-three/fiber';
import { BoxGeometry, Mesh, MeshBasicMaterial, Group } from 'three';

import { useStore } from '../../src/store';
import { mockPolaris } from '../__mocks__/MockPolaris';
// eslint-disable-next-line import/order
import { mockR3F } from '../__mocks__/MockR3F';

mockPolaris();
mockR3F();

import { WebGLCanvasManager } from '../../src/components/WebGLCanvasManager';
import { setFeatureConfig } from '../../src/GlobalSettings';
import { COMPOSER_FEATURES } from '../../src';

import Mock = jest.Mock;

// Mock other internal components
jest.mock('../../src/components/three-fiber/EntityGroup', () => 'EntityGroup');
jest.mock('../../src/components/three-fiber/Environment', () => {
  const originalModule = jest.requireActual('../../src/components/three-fiber/Environment');
  return {
    ...originalModule,
    Environment: 'Environment',
  };
});
jest.mock('../../src/components/three-fiber/StatsWindow', () => {
  return {
    StatsWindow: 'StatsWindow',
  };
});
jest.mock('../../src/components/three-fiber/EditorCamera', () => {
  return {
    EditorMainCamera: 'EditorMainCamera',
  };
});
jest.mock('../../src/components/three-fiber/EditorTransformControls', () => {
  return {
    EditorTransformControls: 'EditorTransformControls',
  };
});
jest.mock('../../src/components/three-fiber/ImmersiveViewCamera', () => {
  return {
    ImmersiveViewCamera: 'ImmersiveViewCamera',
  };
});
jest.mock('../../src/components/three-fiber/SceneInfoView', () => {
  return {
    SceneInfoView: 'SceneInfoView',
  };
});
jest.mock('../../src/components/three-fiber/immersive-view/ImmersiveView', () => {
  return {
    ImmersiveView: 'ImmersiveView',
  };
});

// Mock Hooks
jest.mock('@react-three/fiber', () => {
  const originalModule = jest.requireActual('@react-three/fiber');
  return {
    ...originalModule,
    useThree: jest.fn(),
  };
});

describe('WebGLCanvasManagerSnap', () => {
  const body = document.createElement('body');
  const div = document.createElement('div');
  body.appendChild(div);

  const mockThreeStates = {
    gl: {
      domElement: div,
    },
    size: {
      width: 1920,
      height: 1080,
    },
  };

  const sceneDocument = {
    rootNodeRefs: ['testNode'],
    nodeMap: {
      testNode: { childRefs: ['childNode'] },
      childNode: { parentRef: 'testNode', childRefs: ['grandchildNode'] },
      grandchildNode: { parentRef: 'childNode', childRefs: [] },
    },
  };

  const group = new Group();
  const geometry = new BoxGeometry();
  (geometry as any).uuid = 'Geometry';
  const material = new MeshBasicMaterial();
  (material as any).uuid = 'Material';
  const mesh = new Mesh(geometry, material);
  group.name = 'Test';
  (group as any).uuid = 'Forced';
  (mesh as any).uuid = 'Forced2';
  group.add(mesh);

  const baseState: any = {
    getSceneNodeByRef: jest.fn(),
    getSceneProperty: jest.fn(),
    isEditing: jest.fn(),
    document: sceneDocument,
    highlightedSceneNodeRef: 'childNode',
    getObject3DBySceneNodeRef: jest.fn().mockReturnValue(group),
  };

  beforeEach(() => {
    jest.clearAllMocks();

    const mockUseThree = useThree as Mock;
    mockUseThree.mockReturnValue(mockThreeStates);

    setFeatureConfig({});
  });

  it('should render editing correctly without immersive view feature', () => {
    useStore('default').setState(baseState);
    baseState.isEditing.mockReturnValue(true);
    baseState.getSceneNodeByRef.mockReturnValue('childNode');

    const container = renderer.create(<WebGLCanvasManager />);
    expect(container).toMatchSnapshot();
  });

  it('should render viewing correctly without immersive view feature', () => {
    useStore('default').setState(baseState);
    baseState.isEditing.mockReturnValue(false);
    baseState.getSceneNodeByRef.mockReturnValue('childNode');

    const container = renderer.create(<WebGLCanvasManager />);
    expect(container).toMatchSnapshot();
  });

  it('should render correctly with immersive view feature', () => {
    useStore('default').setState(baseState);
    setFeatureConfig({ [COMPOSER_FEATURES.IMMERSIVE_VIEW]: true });

    baseState.isEditing.mockReturnValue(true);
    baseState.getSceneNodeByRef.mockReturnValue('childNode');

    const container = renderer.create(<WebGLCanvasManager />);

    expect(container).toMatchSnapshot();
  });

  it('should render correctly in immersive view feature', () => {
    const baseState: any = {
      cameraControlsType: 'immersive',
      getSceneNodeByRef: jest.fn(),
      getSceneProperty: jest.fn(),
      isEditing: jest.fn(),
      document: sceneDocument,
      highlightedSceneNodeRef: 'childNode',
      getObject3DBySceneNodeRef: jest.fn().mockReturnValue(group),
    };

    useStore('default').setState(baseState);
    setFeatureConfig({ [COMPOSER_FEATURES.IMMERSIVE_VIEW]: true });

    baseState.isEditing.mockReturnValue(true);
    baseState.getSceneNodeByRef.mockReturnValue('childNode');

    const container = renderer.create(<WebGLCanvasManager />);

    expect(container).toMatchSnapshot();
  });
});
