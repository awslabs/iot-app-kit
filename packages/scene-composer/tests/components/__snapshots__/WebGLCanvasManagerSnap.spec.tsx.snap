// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebGLCanvasManagerSnap should render correctly in immersive view feature 1`] = `
Array [
  <EditorMainCamera />,
  <ImmersiveViewCamera />,
  <ImmersiveView />,
  <group
    dispose={null}
    name="ROOT"
  >
    <EntityGroup
      node="childNode"
    />
  </group>,
  <EffectComposer
    autoClear={false}
  >
    <Outline
      blur={true}
      edgeStrength={5}
      selection={
        Array [
          Object {
            "current": Object {
              "geometries": Array [
                Object {
                  "depth": 1,
                  "depthSegments": 1,
                  "height": 1,
                  "heightSegments": 1,
                  "type": "BoxGeometry",
                  "uuid": "Geometry",
                  "width": 1,
                  "widthSegments": 1,
                },
              ],
              "materials": Array [
                Object {
                  "color": 16777215,
                  "colorWrite": true,
                  "depthFunc": 3,
                  "depthTest": true,
                  "depthWrite": true,
                  "reflectivity": 1,
                  "refractionRatio": 0.98,
                  "stencilFail": 7680,
                  "stencilFunc": 519,
                  "stencilFuncMask": 255,
                  "stencilRef": 0,
                  "stencilWrite": false,
                  "stencilWriteMask": 255,
                  "stencilZFail": 7680,
                  "stencilZPass": 7680,
                  "type": "MeshBasicMaterial",
                  "uuid": "Material",
                },
              ],
              "metadata": Object {
                "generator": "Object3D.toJSON",
                "type": "Object",
                "version": 4.5,
              },
              "object": Object {
                "geometry": "Geometry",
                "layers": 1,
                "material": "Material",
                "matrix": Array [
                  1,
                  0,
                  0,
                  0,
                  0,
                  1,
                  0,
                  0,
                  0,
                  0,
                  1,
                  0,
                  0,
                  0,
                  0,
                  1,
                ],
                "type": "Mesh",
                "uuid": "Forced2",
              },
            },
          },
        ]
      }
      visibleEdgeColor={16777215}
    />
  </EffectComposer>,
  <EditorTransformControls />,
  <GizmoHelper
    alignment="bottom-right"
    margin={
      Array [
        72,
        72,
      ]
    }
  >
    <GizmoViewport
      axisColors={
        Array [
          "#000000",
          "#000000",
          "#000000",
        ]
      }
    />
  </GizmoHelper>,
  <SceneInfoView />,
]
`;

exports[`WebGLCanvasManagerSnap should render correctly with immersive view feature 1`] = `
Array [
  <EditorMainCamera />,
  <ImmersiveViewCamera />,
  <ImmersiveView />,
  <group
    dispose={null}
    name="ROOT"
  >
    <EntityGroup
      node="childNode"
    />
  </group>,
  <EffectComposer
    autoClear={false}
  >
    <Outline
      blur={true}
      edgeStrength={5}
      selection={
        Array [
          Object {
            "current": Object {
              "geometries": Array [
                Object {
                  "depth": 1,
                  "depthSegments": 1,
                  "height": 1,
                  "heightSegments": 1,
                  "type": "BoxGeometry",
                  "uuid": "Geometry",
                  "width": 1,
                  "widthSegments": 1,
                },
              ],
              "materials": Array [
                Object {
                  "color": 16777215,
                  "colorWrite": true,
                  "depthFunc": 3,
                  "depthTest": true,
                  "depthWrite": true,
                  "reflectivity": 1,
                  "refractionRatio": 0.98,
                  "stencilFail": 7680,
                  "stencilFunc": 519,
                  "stencilFuncMask": 255,
                  "stencilRef": 0,
                  "stencilWrite": false,
                  "stencilWriteMask": 255,
                  "stencilZFail": 7680,
                  "stencilZPass": 7680,
                  "type": "MeshBasicMaterial",
                  "uuid": "Material",
                },
              ],
              "metadata": Object {
                "generator": "Object3D.toJSON",
                "type": "Object",
                "version": 4.5,
              },
              "object": Object {
                "geometry": "Geometry",
                "layers": 1,
                "material": "Material",
                "matrix": Array [
                  1,
                  0,
                  0,
                  0,
                  0,
                  1,
                  0,
                  0,
                  0,
                  0,
                  1,
                  0,
                  0,
                  0,
                  0,
                  1,
                ],
                "type": "Mesh",
                "uuid": "Forced2",
              },
            },
          },
        ]
      }
      visibleEdgeColor={16777215}
    />
  </EffectComposer>,
  <EditorTransformControls />,
  <GizmoHelper
    alignment="bottom-right"
    margin={
      Array [
        72,
        72,
      ]
    }
  >
    <GizmoViewport
      axisColors={
        Array [
          "#000000",
          "#000000",
          "#000000",
        ]
      }
    />
  </GizmoHelper>,
  <gridHelper
    args={
      Array [
        1000,
        500,
        0,
        0,
      ]
    }
  />,
  <mesh
    name="Ground"
    onPointerDown={[Function]}
    onPointerMove={[Function]}
    onPointerUp={[Function]}
    rotation={
      Array [
        4.71238898038469,
        0,
        0,
      ]
    }
  >
    <planeGeometry
      args={
        Array [
          1000,
          1000,
        ]
      }
    />
    <meshBasicMaterial
      colorWrite={false}
    />
  </mesh>,
  <SceneInfoView />,
]
`;

exports[`WebGLCanvasManagerSnap should render editing correctly without immersive view feature 1`] = `
Array [
  <EditorMainCamera />,
  <ImmersiveView />,
  <group
    dispose={null}
    name="ROOT"
  >
    <EntityGroup
      node="childNode"
    />
  </group>,
  <EffectComposer
    autoClear={false}
  >
    <Outline
      blur={true}
      edgeStrength={5}
      selection={
        Array [
          Object {
            "current": Object {
              "geometries": Array [
                Object {
                  "depth": 1,
                  "depthSegments": 1,
                  "height": 1,
                  "heightSegments": 1,
                  "type": "BoxGeometry",
                  "uuid": "Geometry",
                  "width": 1,
                  "widthSegments": 1,
                },
              ],
              "materials": Array [
                Object {
                  "color": 16777215,
                  "colorWrite": true,
                  "depthFunc": 3,
                  "depthTest": true,
                  "depthWrite": true,
                  "reflectivity": 1,
                  "refractionRatio": 0.98,
                  "stencilFail": 7680,
                  "stencilFunc": 519,
                  "stencilFuncMask": 255,
                  "stencilRef": 0,
                  "stencilWrite": false,
                  "stencilWriteMask": 255,
                  "stencilZFail": 7680,
                  "stencilZPass": 7680,
                  "type": "MeshBasicMaterial",
                  "uuid": "Material",
                },
              ],
              "metadata": Object {
                "generator": "Object3D.toJSON",
                "type": "Object",
                "version": 4.5,
              },
              "object": Object {
                "geometry": "Geometry",
                "layers": 1,
                "material": "Material",
                "matrix": Array [
                  1,
                  0,
                  0,
                  0,
                  0,
                  1,
                  0,
                  0,
                  0,
                  0,
                  1,
                  0,
                  0,
                  0,
                  0,
                  1,
                ],
                "type": "Mesh",
                "uuid": "Forced2",
              },
            },
          },
        ]
      }
      visibleEdgeColor={16777215}
    />
  </EffectComposer>,
  <EditorTransformControls />,
  <GizmoHelper
    alignment="bottom-right"
    margin={
      Array [
        72,
        72,
      ]
    }
  >
    <GizmoViewport
      axisColors={
        Array [
          "#000000",
          "#000000",
          "#000000",
        ]
      }
    />
  </GizmoHelper>,
  <gridHelper
    args={
      Array [
        1000,
        500,
        0,
        0,
      ]
    }
  />,
  <mesh
    name="Ground"
    onPointerDown={[Function]}
    onPointerMove={[Function]}
    onPointerUp={[Function]}
    rotation={
      Array [
        4.71238898038469,
        0,
        0,
      ]
    }
  >
    <planeGeometry
      args={
        Array [
          1000,
          1000,
        ]
      }
    />
    <meshBasicMaterial
      colorWrite={false}
    />
  </mesh>,
  <SceneInfoView />,
]
`;

exports[`WebGLCanvasManagerSnap should render viewing correctly without immersive view feature 1`] = `
Array [
  <EditorMainCamera />,
  <ImmersiveView />,
  <group
    dispose={null}
    name="ROOT"
  >
    <EntityGroup
      node="childNode"
    />
  </group>,
  <EffectComposer
    autoClear={false}
  >
    <Outline
      blur={true}
      edgeStrength={5}
      selection={
        Array [
          Object {
            "current": Object {
              "geometries": Array [
                Object {
                  "depth": 1,
                  "depthSegments": 1,
                  "height": 1,
                  "heightSegments": 1,
                  "type": "BoxGeometry",
                  "uuid": "Geometry",
                  "width": 1,
                  "widthSegments": 1,
                },
              ],
              "materials": Array [
                Object {
                  "color": 16777215,
                  "colorWrite": true,
                  "depthFunc": 3,
                  "depthTest": true,
                  "depthWrite": true,
                  "reflectivity": 1,
                  "refractionRatio": 0.98,
                  "stencilFail": 7680,
                  "stencilFunc": 519,
                  "stencilFuncMask": 255,
                  "stencilRef": 0,
                  "stencilWrite": false,
                  "stencilWriteMask": 255,
                  "stencilZFail": 7680,
                  "stencilZPass": 7680,
                  "type": "MeshBasicMaterial",
                  "uuid": "Material",
                },
              ],
              "metadata": Object {
                "generator": "Object3D.toJSON",
                "type": "Object",
                "version": 4.5,
              },
              "object": Object {
                "geometry": "Geometry",
                "layers": 1,
                "material": "Material",
                "matrix": Array [
                  1,
                  0,
                  0,
                  0,
                  0,
                  1,
                  0,
                  0,
                  0,
                  0,
                  1,
                  0,
                  0,
                  0,
                  0,
                  1,
                ],
                "type": "Mesh",
                "uuid": "Forced2",
              },
            },
          },
        ]
      }
      visibleEdgeColor={16777215}
    />
  </EffectComposer>,
]
`;
